### PR TITLE
Video Player v0.2

### PR DESCRIPTION
**Pull request by konopka90**
Additional dependency:
   - FFmpeg library

Features:
   - Eliminated mem leaks and seg faults 
   - Enhanced code clarity
   - Custom video width and height!

To do:
   - Audio
   - Read FPS from file (on the other hand all videos in MPQ has 15FPS)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -85,6 +87,8 @@ link_libraries (
     ${PYTHON_LIBRARIES}
     ${Boost_LIBRARIES}
     ${Qt5Widgets_LIBRARIES}
+    ${FFMPEG_LIBRARIES}
+    swresample
```
Comment (wheybags) [inline CMakeLists.txt]: this is unnecessary, and bypasses cmake

File: components/CMakeLists.txt [hunk 2]
```diff
@@ -85,3 +85,10 @@ add_library(Settings
 )
 target_link_libraries(Settings ${Boost_Libraries})
 set_target_properties(Settings PROPERTIES COMPILE_FLAGS "${FA_COMPILER_FLAGS}")
+
+add_library(Video
+    video/video.h
+    video/video.cpp
+)
+target_link_libraries(Video ${FFMPEG_LIBRARIES} swresample)
```
Comment (wheybags) [inline components/CMakeLists.txt]: same as before

File: components/video/video.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+          }
+
+        // Free the packet that was allocated by av_read_frame
+        av_free_packet(&packet);
+    }
+
+    av_free(bufferRGB);
+
+    avformat_close_input(&formatContext);
+    avformat_free_context(formatContext);
```
Comment (wheybags) [inline components/video/video.cpp]: avformat_close_input frees this already, and sets the pointer to NULL

File: components/video/video.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+      mNumFrames(0)
+{
+}
+
+Video::~Video()
+{
+    int size = mSprites.size();
+    for(int i = 0 ; i < size; i++)
+    {
+        SDL_DestroyTexture(reinterpret_cast<SDL_Texture*>(mSprites[i]));
```
Comment (wheybags) [inline components/video/video.cpp]: This should be handled in the render component, nothing but that should know sdl even exists

**Comment by wheybags** [comment]
Had to add 
```
#ifndef AV_CODEC_ID_SMACKVIDEO
    #define AV_CODEC_ID_SMACKVIDEO CODEC_ID_SMACKVIDEO
#endif

#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(55,28,1)
    #define av_frame_alloc  avcodec_alloc_frame
#endif
```
to get it to build on my system, but apart from that, and the few issues I mentioned above it looks great!

Would you like me to merge now, or do you want to take care of the to do items first?
Also, sorry it took so long to get to this, haven't had the time to look at it properly until now.

**Comment by konopka90** [comment]
I'm trying to do better implementation - this solution consumes a lot of memory so please don't merge now. Loading times in FullHD are confusing (all frames are loaded at once). I was trying to do little hack: load frames in small resolution and then scale it in renderer but scaling in renderer doesn't look good - it should happen in "sws_scale" function with proper filter (e.g. bilinear or bicubic). I think about typical producer-consumer problem: producer produces frames in the fly and consumer displays these frames.

**Comment by wheybags** [comment]
Yeah, I was thinking you only need one SDL_texture, which you can just
update every frame, but this would do for now

File: CMakeLists.txt (hunk 1 above)
Comment (konopka90) [inline CMakeLists.txt]: I remember that I had problem with linking on Ubuntu without these lines but I'm not sure for now, I'll check it.

**Pull request closed without merging**